### PR TITLE
Inital Devcontainer config files

### PR DESCRIPTION
**Pull request by GavinRay97**
This PR adds the ability to build & develop Seastar using VS Code Devcontainers, either in the browser via Codespaces, or locally.

It's based off of the VS Code `C++ Devcontainer` template (Ubuntu 22.04 variant selected) the only modifications are:
- Setting `"CMAKE_CXX_COMPILER_LAUNCHER": "ccache"` to use `ccache` in the default VS Code settings
- Automatically running `sudo apt-get update && sudo apt-get install -y ccache && sudo ./install-dependencies.sh` during container post-create to set up deps + `ccache`

It comes with LLVM 14 and GCC 11.

I tested building seastar with it and it works fantastically, no user interaction required beyond just "Open in Codespaces" or "Open in Development Container" + clicking the "build" button 👍 

Intend to modify existing documentation in a follow up PR

### CODE REVIEW

**Review by nyh — CHANGES_REQUESTED** [review]
Thanks, I left a few comments/questions, but because I've never even heard of "vscode devcontainer" until a few minutes ago, some of these questions might be stupid (in that case, sorry).

Please note that in the way this project works, when the PR has a single patch its "cover letter" (the first message in the PR) doesn't reach the git history (because the PR is rebased, not merged) - so everything you wrote there to explain the rationale of this patch will not be saved. So please add to the commit message of your single commit an explanation why you are doing this - not a video but rather a few paragraph explaining why "devcontainer" (which again, I admit to being completely ignorant of) is so useful, popular, or whatever, that we really want this file in.

Please also add a file .devcontainer/README explaining what this directory is for - e.g., explaining very shortly what "devcontainer" is and giving a link for more information. A few years into the future users might run into this directory and have no idea what it's for.

Thanks.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ARG VARIANT="bullseye"
+FROM mcr.microsoft.com/vscode/devcontainers/cpp:0-${VARIANT}
+
+# [Optional] Install CMake version different from what base image has already installed. 
+# CMake reinstall choices: none, 3.21.5, 3.22.2, or versions from https://cmake.org/download/
+ARG REINSTALL_CMAKE_VERSION_FROM_SOURCE="none"
+
+# Optionally install the cmake for vcpkg
+COPY ./reinstall-cmake.sh /tmp/
+RUN if [ "${REINSTALL_CMAKE_VERSION_FROM_SOURCE}" != "none" ]; then \
```
[the remark this answers is not in the document]
Comment (nyh) [reply .devcontainer/Dockerfile:13]: If I understand correctly below, it seems a different setting of REINSTALL_CMAKE_VERSION_FROM_SOURCE overrides the one defined in this file, which makes this code relevant. I think it would be clearer if this file had the right setting right from the start, instead of having a bogus setting which is overriden in a separate file.

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/tree/v0.245.2/containers/cpp/.devcontainer/base.Dockerfile
+
+# [Choice] Debian / Ubuntu version (use Debian 11, Ubuntu 18.04/22.04 on local arm64/Apple Silicon): debian-11, debian-10, ubuntu-22.04, ubuntu-20.04, ubuntu-18.04
```
[the remark this answers is not in the document]
Comment (GavinRay97) [reply .devcontainer/Dockerfile:3]: So this file is automatically generated from the `C++ Development` template (no modifications/manual adjustments were made):

https://user-images.githubusercontent.com/26604994/198893800-5dffa8d8-0f3f-4876-82b1-e431e2fb1e90.mp4

Your options for image variants are:
- `debian-11, debian-10, ubuntu-22.04, ubuntu-20.04, ubuntu-18.04`

File: .devcontainer/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (GavinRay97) [reply .devcontainer/Dockerfile:13]: > If I understand correctly, this code doesn't do anything if REINSTALL_CMAKE_VERSION_FROM_SOURCE is "none", which it is (a couple of lines above), so why do you need this code at all, or the "reinstall-cmake.sh" file?

> it seems a different setting of REINSTALL_CMAKE_VERSION_FROM_SOURCE overrides the one defined in this file, which makes this code relevant. I think it would be clearer if this file had the right setting right from the start, instead of having a bogus setting which is overriden in a separate file.

The setting in `ARG` is a default if none is specified, it is part of the base Dockerfile template

The `devcontainer.json` allows users to configure `ARG` parameters to override things like:
- The base OS image tag (ubuntu/debian + version)
- Any other `ARG` parameters in the scripts

This is for two reasons:
- It allows Docker to more intelligently cache layers/changes and prevent spurious rebuilds
- It allows users to override parameters in the devcontainer config rather than needing to modify the Dockerfile

File: .devcontainer/devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.245.2/containers/cpp
+{
+    "name": "C++",
+    "build": {
+        "dockerfile": "Dockerfile",
+        // Update 'VARIANT' to pick an Debian / Ubuntu OS version: debian-11, debian-10, ubuntu-22.04, ubuntu-20.04, ubuntu-18.04
+        // Use Debian 11, Ubuntu 18.04 or Ubuntu 22.04 on local arm64/Apple Silicon
+        "args": {
+            "VARIANT": "ubuntu-22.04",
```
[the remark this answers is not in the document]
Comment (GavinRay97) [reply .devcontainer/devcontainer.json:10]: `ARG`'s are user-providable arguments in Dockerfiles, you can give them during builds like:

```sh
docker build -t essearch/ess-elasticsearch:1.7.6 \
--build-arg number_of_shards=5 \
--build-arg number_of_replicas=2
```

When an `ARG` is declared, it can be given a default value

The values provided in the `devcontainer.json` -> `build.args` are passed to the Docker build API and take precedence over default values

File: .devcontainer/devcontainer.json [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            "REINSTALL_CMAKE_VERSION_FROM_SOURCE": "3.24.2"
```
[the remark this answers is not in the document]
Comment (GavinRay97) [reply .devcontainer/devcontainer.json:11]: > Wouldn't everything be clearer if Dockerfile had the right options instead of being meaningless and overriden by the options in this file?

If this version was hardcoded then (somewhat obviously) you would not be able to change the CMake version in the image, which sort of plays into your second question:

> Also, I'm curious why you need to "reinstall cmake from source" - what's wrong with whatever version of cmake version Ubuntu 22.04 has?

Technically as of the current state of the code, nothing. But say that next year, Seastar adds a CMake change that depends on recent CMake feature. In that case, I figured it would be better to future-proof the change and make it user-configurable + obvious how to set the version.

I'm not married to the idea, happy to revert it if you think something like this will never happen (or just set `REINSTALL_CMAKE_VERSION_FROM_SOURCE=none`)

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    },
+    "runArgs": [
+        "--cap-add=SYS_PTRACE",
+        "--security-opt",
+        "seccomp=unconfined"
```
[the remark this answers is not in the document]
Comment (GavinRay97) [reply .devcontainer/devcontainer.json:17]: Ah, good question

So you don't need them to _build_ CMake, but this is intended to be a _development_ environment (one nice side effect of which is easily building)

(See long comment about Devcontainers in general posted as reply to original PR)

These flags allow you to properly debug inside of the container. For example, `--security-opt seccomp=unconfined` allows use of `gdb`/`lldb`.

Without it, you'll get: `Error disabling address space randomization: Operation not permitted`, and the debugger will fail

**Comment by GavinRay97** [comment]
@nyh I should have given a summary of a what Devcontainers are in the original PR description, that was my fault.

Hopefully the below clears up some of the stuff in the PR that might seem strange/raise eyebrows:

---

- "Devcontainers" are an OSS specification for tooling to work with developing software inside of containerized environments (Docker/OCI images)
	- https://containers.dev/
	- https://github.com/devcontainers/spec
	- [VSCode - Developing Inside a Container](https://code.visualstudio.com/docs/devcontainers/containers)
- tl;dr = they:
	- mount your local filesystem into the container
	- handle things like port forwarding/debugging capabilities to ensure it works properly & transparently
	- which allows you to work from your host OS, while using a toolset and dependencies running inside of a container.

**Comment by nyh** [comment]
Thanks. I wasn't familiar with https://github.com/devcontainers/spec. Interestingly, we have had in the ScyllaDB project (the open-source database built on Seastar) a similar concept, which we call "dbuild" for a long time - it's a frozen toolset kept as a docker image, which can be used to build the project with exactly the same compilers/libraries/etc. on every build host.

**Comment by GavinRay97** [comment]
Just an update -- I intend to finish this PR + adopt the suggested changes, was travelling this past week and then back at work.
This weekend I should have time to get this done.

One question though:

> "So please add to the commit message of your single commit an explanation why you are doing this"

How do I do this -- modify an existing commit message without screwing up the history?
Should I finish the rest of the PR/commits, and then squash them into a single commit with the intended message?
